_posts/2025-08-06-spatial-joins.md

```diff
@@ -0,0 +1,336 @@
+
+Ever since its inception, DuckDB's `spatial` extension has provided a `GEOMETRY` column type to represent locations, regions and shapes, along with plenty of _spatial predicate_ functions to use when `JOIN`:ing. But it wasn't until recently, in DuckDB v1.3.0 that spatial joins really became _scalable_ thanks to the introduction of the dedicated `SPATIAL_JOIN` query **operator**. 
+
+I managed to hastily sneak in some information about it towards [the end of the v1.3.0 release notes]({% post_url 2025-05-21-announcing-duckdb-130 %}#spatial-join-operator), but we figured there is enough say about it to justify a separate blog post. So in this post, we will take a closer look at some of the challenges of optimizing spatial joins in DuckDB and how the new `SPATIAL_JOIN` operator achieves new levels of efficency.
+
+## What does spatial join look like again?
+
+Spatial joins are incredibly important in geospatial analysis, and performing one in SQL is actually surprisingly easy. They don't require any special syntax or magic incantation, you just have to join two tables with a `GEOMETRY` column using some _spatial predicate_. Below is a simple query that joins two tables `a` and `b` on the condition that the geometries in `a.geom` and `b.geom` “intersect”, using the `ST_Intersects` spatial predicate function:
+
+```sql
+SELECT * FROM a JOIN b ON ST_Intersects(a.geom, b.geom)
+```
+
+Let's try to do something more advanced to demonstrate. We're going to analyze the [New York City Citi Bike Trip dataset](https://citibikenyc.com/system-data), which contains around 58 million rows representing rental bike trips in New York City, including the start and end locations of each trip. We want to find neighborhoods in NYC that have the most bike trips starting in them. To do this, we will join the bike trip data with a [dataset of NYC neighborhoods](https://www.nyc.gov/content/planning/pages/resources/datasets/neighborhood-tabulation), count and group the number of trips starting in each neighborhood, and finally sort a limit to return the top 3 groups. We've compiled these datasets and extracted the relevant columns for this example into two tables, `rides` and `hoods` in a [DuckDB database available to download here](https://duckdb-blobs.s3.us-east-1.amazonaws.com/biketrips.db)
+
+The query looks like this:
+
+```sql
+LOAD spatial; -- Load the spatial extension
+
+SELECT neighborhood, count(*) as num_rides 
+FROM rides JOIN hoods on st_intersects(rides.start_geom, hoods.geom) 
+GROUP BY neighborhood 
+ORDER BY num_rides 
+LIMIT 3;
+```
+
+```text
+┌──────────────┬───────────┐
+│ neighborhood │ num_rides │
+│   varchar    │   int64   │
+├──────────────┼───────────┤
+│ Midtown      │   6253835 │
+│ Chelsea      │   6238693 │
+│ East Village │   4047396 │
```

Comment:
Wasn't there a way to add undrescores? Just wondering if that'd be cool. `6_253_835`

Reply:
No, only when parsing numbers I think.